ccvm_simulators/solvers/dl_solver.py

```diff
@@ -444,3 +470,288 @@ def solve(
             solution.evolution_file = evolution_file
 
         return solution
+
+    def __call__(
+        self,
+        instance,
+        post_processor=None,
+        pump_rate_flag=True,
+        g=0.05,
+        evolution_step_size=None,
+        evolution_file=None,
+        adam_hyperparam=dict(beta1=0.9, beta2=0.999, alpha=0.001),
+    ):
+        """Solves the given problem instance using the DL-CCVM solver including ADAM algorithm.
+
+        Args:
+            instance (ProblemInstance): The problem instance to solve.
+            post_processor (str): The name of the post processor to use to process the results of the solver.
+                None if no post processing is desired. Defaults to None.
+            pump_rate_flag (bool): Whether or not to scale the pump rate based on the
+            iteration number. If False, the pump rate will be 1.0. Defaults to True.
+            g (float): The nonlinearity coefficient. Defaults to 0.05.
+            evolution_step_size (int): If set, the c/s values will be sampled once
+                per number of iterations equivalent to the value of this variable.
+                At the end of the solve process, the best batch of sampled values
+                will be written to a file that can be specified by setting the evolution_file parameter.
+                Defaults to None, meaning no problem variables will be written to the file.
+            evolution_file (str): The file to save the best set of c/s samples to.
+                Only revelant when evolution_step_size is set.
+                If a file already exists with the same name, it will be overwritten.
+                Defaults to None, which generates a filename based on the problem instance name.
+            adam_hyperparam (dict): Hyperparameters for adam algorithm. Defaults to the paper.
+
+        Returns:
+            solution (Solution): The solution to the problem instance.
+        """
+        # If the instance and the solver don't specify the same device type, raise
+        # an error
+        if instance.device != self.device:
+            raise ValueError(
+                f"The device type of the instance ({instance.device}) and the solver"
+                f" ({self.device}) must match."
+            )
+
+        # Get problem from problem instance
+        problem_size = instance.problem_size
+        self.q_matrix = instance.q_matrix
+        self.v_vector = instance.v_vector
```

Comment:
There are couple of issues in the definition of `DLSolver._calculate_grads_boxqp()` and I guess it is the right time to bring them up here. In that way we can decide which method either `DLSolver._calculate_grads_boxqp(...)` or `DLSolver._calculate_grads_boxqp_adam()` to the entire system.
1. Let's start with the signature of this method: `DLSolver_calculate_grads_boxqp(self, c, s, q_matrix, v_vector, pump, rate, S=1)` in which `q_matrix` and `v_vector` are constant arguments and they are passed `iterations=15000` times. In order to avoid unnecessary memory usage repeatedly, we are assigning their values to `self` variables (attributes) in `DLSolver._calculate_grads_boxqp_adam(self, c, s, S=1)`'s  implementation.
2. It seems to me the use of `_calculate_grad_boxqp()` method in existing `solve()` is semantically incorrect  (or inconsistent) because it evaluates the drift part of DL-CCVM:
3.

Reply:
I see, so it sounds like both cases should be using self.q_matrix and self.v_vector! Thanks for the explanation.
In that case, this code is good and fixing the other signature when we address the issues with the drift calculation sounds like a good idea to me!

Reply:
As suggested, 
1. `_calculate_grads_boxqp()` is replaced with `_calculate_drift_boxqp()` and so does `_calculate_grads_boxqp_adam()` is replaced with `_calculate_grads_boxqp()`
2. The signature of drift method is updated in all solver classes with an attribute assignment of Q matrix and V vector. Besides the body of `MFSolver._calculate_drift_boxqp()` has a minor modification to move the diffusion term to the main loop of `MFSolver._solve()`
3. `_method_selector()` is updated and moved to the base class `CCVMSolver` to avoid duplication

Reply:
Thanks, Mehmet. That's perfect!